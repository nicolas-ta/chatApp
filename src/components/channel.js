import React, {useEffect, useState, useMemo} from 'react';
import {Box, Text} from 'native-base';
import {Image, TouchableOpacity} from 'react-native';
// import {} from 'react-native-gesture-handler';
import moment from 'moment';
import channelStyle from '../styles/channel.style';

import {withAppContext} from '../context';
import {
  createChannelName,
  createUnreadMessageCount,
  ellipsis,
} from '../misc/utils';

const LAST_MESSAGE_ELLIPSIS = 23;

const Channel = props => {
  const {sendbird, channel, onPress} = props;
  const [name, setName] = useState('');
  const [lastMessage, setLastMessage] = useState('');
  const [unreadMessageCount, setUnreadMessageCount] = useState('');
  const [updatedAt, setUpdatedAt] = useState('');

  const channelHandler = useMemo(
    () => new sendbird.ChannelHandler(),
    [sendbird.ChannelHandler],
  );
  channelHandler.onChannelChanged = updatedChannel => {
    if (updatedChannel.url === channel.url) {
      updateChannelName(updatedChannel);
      updateLastMessage(updatedChannel);
      updateUnreadMessageCount(updatedChannel);
      updateUpdatedAt(updatedChannel);
    }
  };
  channelHandler.onUserJoined = (updatedChannel, user) => {
    if (updatedChannel.url === channel.url) {
      if (user.userId !== sendbird.currentUser.userId) {
        updateChannelName(updatedChannel);
      }
    }
  };
  channelHandler.onUserLeft = (updatedChannel, user) => {
    if (updatedChannel.url === channel.url) {
      if (user.userId !== sendbird.currentUser.userId) {
        updateChannelName(updatedChannel);
      }
    }
  };

  const updateChannelName = newChan => {
    setName(createChannelName(newChan));
  };

  const updateLastMessage = newChan => {
    if (newChan.lastMessage) {
      const message = newChan.lastMessage;
      if (message.isUserMessage()) {
        setLastMessage(message.message);
      } else if (message.isFileMessage()) {
        setLastMessage(message.name);
      }
    }
  };
  const updateUnreadMessageCount = newChan => {
    setUnreadMessageCount(createUnreadMessageCount(newChan));
  };
  const updateUpdatedAt = newChan => {
    setUpdatedAt(
      moment(
        newChan.lastMessage ? newChan.lastMessage.createdAt : newChan.createdAt,
      ).fromNow(),
    );
  };

  useEffect(() => {
    // channel event listener
    sendbird.addChannelHandler(`channel_${channel.url}`, channelHandler);
    updateChannelName(channel);
    updateLastMessage(channel);
    updateUnreadMessageCount(channel);
    updateUpdatedAt(channel);
    return () => {
      sendbird.removeChannelHandler(`channel_${channel.url}`);
    };
  }, [channel, channelHandler, sendbird]);
  return (
    <TouchableOpacity
      activeOpacity={0.8}
      style={channelStyle.channelContainer}
      onPress={() => onPress(channel)}>
      <Image
        source={
          channel.coverUrl
            ? {uri: channel.coverUrl}
            : require('../assets/logo-icon-purple.png')
        }
        style={channelStyle.profileImage}
      />
      <Box style={channelStyle.contentContainer}>
        <Text style={channelStyle.name}>{name}</Text>
        <Text style={channelStyle.lastMessage}>
          {ellipsis(lastMessage.replace(/\n/g, ' '), LAST_MESSAGE_ELLIPSIS)}
        </Text>
      </Box>
      <Box style={channelStyle.propertyContainer}>
        <Text style={channelStyle.updatedAt}>{updatedAt}</Text>
        {channel.unreadMessageCount > 0 ? (
          <Box style={channelStyle.unreadMessageCountContainer}>
            <Text style={channelStyle.unreadMessageCount}>
              {unreadMessageCount}
            </Text>
          </Box>
        ) : null}
      </Box>
    </TouchableOpacity>
  );
};

export default withAppContext(Channel);
