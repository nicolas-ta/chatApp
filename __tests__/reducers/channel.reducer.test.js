import {channelReducer} from '@reducers/channel';

const channel1 = {
  url: 'url1',
  lastMessage: 'lastMessage1',
  createdAt:
    'Fri Sep 17 2021 15:43:14 GMT+0200 (heure d’été d’Europe centrale)',
};
const channel2 = {
  url: 'url2',
  lastMessage: 'lastMessage2',
  createdAt:
    'Fri Sep 17 2021 15:43:26 GMT+0200 (heure d’été d’Europe centrale)',
};
const channel3 = {
  url: 'url3',
  lastMessage: 'lastMessage3',
  createdAt:
    'Fri Sep 17 2021 15:50:50 GMT+0200 (heure d’été d’Europe centrale)',
};

describe('Channel Reducer unit tests', () => {
  test('checking initial state', () => {
    const action = {type: 'nonexistant'};
    expect(channelReducer(undefined, action)).toMatchSnapshot();
  });

  test('checking refresh state', () => {
    const action = {type: 'refresh'};
    expect(channelReducer(undefined, action)).toMatchSnapshot();
  });

  test('checking fetch-channels empty state', () => {
    const action = {
      type: 'fetch-channels',
    };
    const state = {channels: [channel1], channelMap: {}};
    expect(channelReducer(state, action)).toMatchSnapshot();
  });

  test('checking fetch-channels state', () => {
    const action = {
      type: 'fetch-channels',
      payload: {channels: [channel1, channel2]},
    };
    const state = {channels: [channel1, channel2, channel3], channelMap: {}};
    expect(channelReducer(state, action)).toMatchSnapshot();
  });

  test('checking fetch-open-channels state', () => {
    const action = {type: 'fetch-open-channels'};
    const state = {openChannels: []};
    expect(channelReducer(state, action)).toMatchSnapshot();
  });

  test('checking fetch-open-channels with channels state', () => {
    const action = {
      type: 'fetch-open-channels',
      payload: {channels: [channel1, channel2]},
    };
    const state = {
      openChannels: [channel1, channel2],
      openChannelMap: {channel2},
    };
    expect(channelReducer(state, action)).toMatchSnapshot();
  });

  test('checking join-channel state', () => {
    const action = {type: 'join-channel'};
    const state = {channelMap: [], channels: []};

    expect(channelReducer(state, action)).toMatchSnapshot();
  });

  test('checking join-channel with channel state', () => {
    const action = {type: 'join-channel', payload: {channel: channel1}};
    const state = {
      channelMap: {channel1},
      channels: [channel1, channel2],
    };
    expect(channelReducer(state, action)).toMatchSnapshot();
  });

  test('checking update-channel state', () => {
    const action = {type: 'update-channel'};
    const state = {
      channelMap: {channel1},
      channels: [channel1, channel2],
    };

    expect(channelReducer(state, action)).toMatchSnapshot();
  });

  test('checking update-channel with channel state', () => {
    const action = {type: 'update-channel', payload: {channel: channel1}};
    const state = {
      channelMap: {channel1},
      channels: [channel1, channel2],
    };

    expect(channelReducer(state, action)).toMatchSnapshot();
  });

  test('checking leave-channel state', () => {
    const action = {type: 'leave-channel'};
    const state = {channels: []};

    expect(channelReducer(state, action)).toMatchSnapshot();
  });

  test('checking delete-channel state', () => {
    const action = {type: 'delete-channel', payload: {channel: channel1}};
    const state = {channels: [channel1, channel2]};

    expect(channelReducer(state, action)).toMatchSnapshot();
  });

  test('checking start-loading state', () => {
    const action = {type: 'start-loading'};
    expect(channelReducer(undefined, action)).toMatchSnapshot();
  });

  test('checking end-loading state', () => {
    const action = {type: 'end-loading'};
    expect(channelReducer(undefined, action)).toMatchSnapshot();
  });

  test('checking error state', () => {
    const action = {type: 'error'};
    expect(channelReducer(undefined, action)).toMatchSnapshot();
  });
});
