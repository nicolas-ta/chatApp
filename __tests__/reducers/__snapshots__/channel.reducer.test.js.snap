// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Channel Reducer unit tests checking delete-channel state 1`] = `
Object {
  "channelMap": Object {
    "url1": false,
  },
  "channels": Array [
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:26 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage2",
      "url": "url2",
    },
  ],
  "empty": "",
}
`;

exports[`Channel Reducer unit tests checking end-loading state 1`] = `
Object {
  "error": null,
  "loading": false,
}
`;

exports[`Channel Reducer unit tests checking error state 1`] = `
Object {
  "error": undefined,
}
`;

exports[`Channel Reducer unit tests checking fetch-channels state 1`] = `
Object {
  "channelMap": Object {
    "url1": true,
    "url2": true,
    "url3": true,
  },
  "channels": Array [
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:14 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage1",
      "url": "url1",
    },
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:26 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage2",
      "url": "url2",
    },
    Object {
      "createdAt": "Fri Sep 17 2021 15:50:50 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage3",
      "url": "url3",
    },
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:14 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage1",
      "url": "url1",
    },
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:26 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage2",
      "url": "url2",
    },
  ],
  "empty": "",
}
`;

exports[`Channel Reducer unit tests checking fetch-open-channels state 1`] = `
Object {
  "channelMap": Object {},
  "empty": "No conversation",
  "openChannels": Array [],
}
`;

exports[`Channel Reducer unit tests checking fetch-open-channels with channels state 1`] = `
Object {
  "channelMap": Object {
    "url1": true,
    "url2": true,
  },
  "empty": "",
  "openChannelMap": Object {
    "channel2": Object {
      "createdAt": "Fri Sep 17 2021 15:43:26 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage2",
      "url": "url2",
    },
  },
  "openChannels": Array [
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:14 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage1",
      "url": "url1",
    },
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:26 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage2",
      "url": "url2",
    },
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:14 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage1",
      "url": "url1",
    },
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:26 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage2",
      "url": "url2",
    },
  ],
}
`;

exports[`Channel Reducer unit tests checking initial state 1`] = `undefined`;

exports[`Channel Reducer unit tests checking join-channel state 1`] = `
Object {
  "channelMap": Object {},
  "channels": Array [],
  "empty": "",
}
`;

exports[`Channel Reducer unit tests checking join-channel with channel state 1`] = `
Object {
  "channelMap": Object {
    "channel1": Object {
      "createdAt": "Fri Sep 17 2021 15:43:14 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage1",
      "url": "url1",
    },
    "url1": true,
  },
  "channels": Array [
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:14 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage1",
      "url": "url1",
    },
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:26 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage2",
      "url": "url2",
    },
  ],
  "empty": "",
}
`;

exports[`Channel Reducer unit tests checking leave-channel state 1`] = `
Object {
  "channelMap": Object {},
  "channels": Array [],
  "empty": "No conversation",
}
`;

exports[`Channel Reducer unit tests checking refresh state 1`] = `
Object {
  "channelMap": Object {},
  "channels": Array [],
  "error": null,
  "loading": false,
  "openChannelMap": Object {},
  "openChannels": Array [],
}
`;

exports[`Channel Reducer unit tests checking start-loading state 1`] = `
Object {
  "error": null,
  "loading": true,
}
`;

exports[`Channel Reducer unit tests checking update-channel state 1`] = `
Object {
  "channelMap": Object {},
  "channels": Array [],
  "empty": "",
}
`;

exports[`Channel Reducer unit tests checking update-channel with channel state 1`] = `
Object {
  "channelMap": Object {
    "channel1": Object {
      "createdAt": "Fri Sep 17 2021 15:43:14 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage1",
      "url": "url1",
    },
    "url1": true,
  },
  "channels": Array [
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:14 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage1",
      "url": "url1",
    },
    Object {
      "createdAt": "Fri Sep 17 2021 15:43:26 GMT+0200 (heure d’été d’Europe centrale)",
      "lastMessage": "lastMessage2",
      "url": "url2",
    },
  ],
  "empty": "",
}
`;
